internal/auth/auth.go

```diff
@@ -0,0 +1,173 @@
+	return bip32.GetPrivateKeyFromHDKey(derivedKey)
+}
+
+// Generate unlocking script using private key
+func getUnlockingScript(privateKey *ec.PrivateKey) (*p2pkh.P2PKH, error) {
+	sigHashFlags := sighash.AllForkID
+	return p2pkh.Unlock(privateKey, &sigHashFlags)
+}
+
+// createSignature will create a signature for the given key & body contents
+func createSignature(xPriv *bip32.ExtendedKey, bodyString string) (payload *models.AuthPayload, err error) {
+	// No key?
+	if xPriv == nil {
+		err = ErrMissingXpriv
+		return
+	}
+
+	// Get the xPub
+	payload = new(models.AuthPayload)
+	if payload.XPub, err = bip32.GetExtendedPublicKey(xPriv); err != nil { // Should never error if key is correct
+		return
+	}
+
+	// auth_nonce is a random unique string to seed the signing message
+	// this can be checked server side to make sure the request is not being replayed
+	if payload.AuthNonce, err = cryptoutil.RandomHex(32); err != nil { // Should never error if key is correct
+		return
+	}
+
+	// Derive the address for signing
+	var key *bip32.ExtendedKey
+	if key, err = cryptoutil.DeriveChildKeyFromHex(xPriv, payload.AuthNonce); err != nil {
+		return
+	}
+
+	var privateKey *ec.PrivateKey
+	if privateKey, err = bip32.GetPrivateKeyFromHDKey(key); err != nil {
+		return // Should never error if key is correct
+	}
+	return createSignatureCommon(payload, bodyString, privateKey)
+}
+
+// createSignatureCommon will create a signature
+func createSignatureCommon(payload *models.AuthPayload, bodyString string, privateKey *ec.PrivateKey) (*models.AuthPayload, error) {
+	// Create the auth header hash
+	payload.AuthHash = cryptoutil.Hash(bodyString)
+	// auth_time is the current time and makes sure a request can not be sent after 30 secs
+	payload.AuthTime = time.Now().UnixMilli()
+
+	key := payload.XPub
+	if key == "" && payload.AccessKey != "" {
+		key = payload.AccessKey
+	}
+	// Signature, using bitcoin signMessage
+	sigBytes, err := bsm.SignMessage(privateKey, getSigningMessage(key, payload))
+	if err != nil {
+		return nil, err
+	}
+
+	payload.Signature = base64.StdEncoding.EncodeToString(sigBytes)
+	return payload, nil
+}
+
+// getSigningMessage will build the signing message byte array
+func getSigningMessage(xPub string, auth *models.AuthPayload) []byte {
+	message := fmt.Sprintf("%s%s%s%d", xPub, auth.AuthHash, auth.AuthNonce, auth.AuthTime)
+	return []byte(message)
+}
+
+func setSignatureHeaders(header *http.Header, authData *models.AuthPayload) {
+	// Create the auth header hash
+	header.Set(models.AuthHeaderHash, authData.AuthHash)
+	// Set the nonce
+	header.Set(models.AuthHeaderNonce, authData.AuthNonce)
+	// Set the time
+	header.Set(models.AuthHeaderTime, fmt.Sprintf("%d", authData.AuthTime))
+	// Set the signature
+	header.Set(models.AuthSignature, authData.Signature)
+}
+
+// ErrMissingXpriv is when xpriv is missing
+var ErrMissingXpriv = models.SPVError{Message: "xpriv is missing", StatusCode: 401, Code: "error-unauthorized-xpriv-missing"}
```

Comment:
@chris-4chain I didn't understand the part about errors, so can you look at this and give your opinion

Reply:
Generally, return SPVError only if it was returned by `spv-wallet`. (I mean a message with a code)